### PR TITLE
Skyline: Handle latest chromatogram library export from Panorama, inc…

### PR DESCRIPTION
**Pull request by bspratt**
…luding new support for ion mobility and small molecules.

### CODE REVIEW

File: pwiz_tools/Skyline/Model/Lib/ChromLib/ChromatogramLibrary.cs [hunk 1]
```diff
@@ -91,19 +92,35 @@ protected override LibraryChromGroup ReadChromatogram(ChromLibSpectrumInfo info)
...
+                            {
+                                ionMobility = IonMobilityValue.GetIonMobilityValue(im, imUnits);
+                            }
+                        }
                         chromDatas.Add(new LibraryChromGroup.ChromData
                             {
                                 Mz = transition.Mz,
                                 Height = transition.Height,
                                 Intensities = timeIntensities.Intensities[transition.ChromatogramIndex],
-                                Charge = transition.Charge == 0 ? precursor.Charge : transition.Charge,
```
Comment (nickshulman) [inline pwiz_tools/Skyline/Model/Lib/ChromLib/ChromatogramLibrary.cs:101]: I think the removal of this line is causing the SplitGraphTest failure.
In older .clib files, the Charge column is sometimes Null for some transitions, in which case it's supposed to use the charge from the precursor.

File: pwiz_tools/Skyline/Model/Lib/ChromLib/ChromatogramLibrary.cs [hunk 2]
```diff
@@ -249,12 +298,46 @@ private bool LoadLibraryFromDatabase(ILoadMonitor loader)
...
+                    var dictMolecules = new Dictionary<int, CustomMolecule>();
+                    var dictMoleculeLists = new Dictionary<int, Protein>();
+                    if (hasSmallMoleculAndIonMobilityColumns)
+                    {
+                        var moleculeQuery = session.CreateQuery(@"SELECT Id FROM Peptide"); 
+                        foreach (var peptideId in moleculeQuery.List<int>())
+                        {
+                            if (dictMolecules.ContainsKey(peptideId))
+                            {
+                                continue; // No idea while the query runs this list twice
```
[the remark this answers is not in the document]
Comment (bspratt) [reply pwiz_tools/Skyline/Model/Lib/ChromLib/ChromatogramLibrary.cs:311]: > Is there a typo in this comment? I can't figure out what it's supposed to be saying.

Yes, that should be "why" not "while". But it does indeed seem to iterate over that list twice. Any idea why?

File: pwiz_tools/Skyline/Model/Lib/ChromLib/ChromatogramLibrary.cs (hunk 1 above)
Comment (bspratt) [reply pwiz_tools/Skyline/Model/Lib/ChromLib/ChromatogramLibrary.cs:101]: Thanks!

File: pwiz_tools/Skyline/Model/Lib/ChromLib/ChromatogramLibrary.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (nickshulman) [reply pwiz_tools/Skyline/Model/Lib/ChromLib/ChromatogramLibrary.cs:311]: This has to do with Hibernate and polymorphism. Vagisha figured this one out back in 2017. The fix is to add:
polymorphism="explicit"
to something in the mapping.xml file.

I'll forward you Vagisha's email that she sent me back in 2017.

**Comment by bspratt** [comment]
Here's a view of a mixed proteomic/smallMol clib with ion mobility info in Library Explorer.
Note that we now show that in the tooltip when available, and that we now show the file and IM/CCS in the main pane as we already did with spectral libraries.

image

**Comment by brendanx67** [comment]
One strange thing about the chromatogram view you are showing in the screenshot above, is that there is nowhere that you can find the fragment m/z value for the chromatogram. I assume that you can still switch back to spectrum view to see it, and that the RT would then be visible in the file name line as well?

**Comment by bspratt** [comment]
Thanks - must be some peptide-specific filtering going on there. I expect it's an easy fix.

**Comment by bspratt** [comment]
Hmm. Near as I can tell, it's acting exactly the same way Skyline-Daily does. I just didn't have everything turned on, I guess:
image

>>  I assume that you can still switch back to spectrum view to see it, and that the RT would then be visible in the file name line as well?
I'm not sure what RT means in the context of a chromatogram library entry? 

FWIW here are some screenshots of Daily showing a proteomic clib:
image
image
Are these as you would want them?

**Comment by brendanx67** [comment]
Not really. What happened to the y- and b-ion annotations in the spectrum plot? Also the line breaks in the legend look pretty awful. And, I think you want to show the "Ion m/z values" in the legend on the chromatogram plot regardless, because it is not included as the x-axis, as in the spectrum plot. I see how the checked menu items I developed for the spectrum plots might sort of give you something for "free" in the chromatogram plot, but they need to be rethought in this new context, considering line breaks and whether they still really make sense.

**Comment by bspratt** [comment]
I agree with all that. But it should be made clear that this is Skyline's current behavior, none of my changes for IMS and small molecule in CLIB are in the code that produced the two images showing the popup menus. Do we want to fix that as part of my branch, or on its own?

**Comment by bspratt** [comment]
>> I'm not sure what RT means in the context of a chromatogram library entry?

I suppose it's just the RT of the chromatogram peak?

**Comment by brendanx67** [comment]
I don't see the annotations missing when I turn of "Show Chromatograms".
image

**Comment by brendanx67** [comment]
But I do see a bunch of other strangeness, we should probably work through separately. I obviously haven't tested this thoroughly enough before. Yes, the RT should be the retention time you see in the Chromatogram, the apex of the integrated peak. That is weird that we don't produce retention times as they are understood by Skyline from this type of library, when we expect the retention times stored in these libraries to be more accurate than the ones stored in spectral libraries from DDA.

**Comment by bspratt** [comment]
>> I don't see the annotations missing when I turn of "Show Chromatograms".

Mysterious. I guess I'd like to understand this before I merge. Can you do a screengrab of the popup menu showing your options?

>>  That is weird that we don't produce retention times as they are understood by Skyline from this type of library

It may be that we're just not showing them here. This seems like a reasonable one to do on this branch.

**Comment by bspratt** [comment]
I find I can get Daily in and out of that state of missing annotations with some sequence of menu selections, so now I just have to figure out what that recipe is then fix it.

**Comment by bspratt** [comment]
OK, this should be good to go. 

BTW I notice that a lot of information we show in Library Explorer isn't shown in Library Match, is that intentional?

**Comment by bspratt** [comment]
Multiple small mol ID types on seperate lines in tooltip:
image

cleaner formatting of IM values, and legend shown regardless of settings used by other views:
image

no "Scores" option in popup menu for chromatograms
image

**Pull request merged**